### ISSUE TITLE
PRIV_TOP conflicts with PRIV_TASK

### ISSUE THREAD

**Issue by Dridi** [varnishcache/varnish-cache#3003]
I was struggling to back-port a bunch of commits in this area to 6.0, eventually decided to test trunk, and I realized the conflict exists there too:

```
varnishtest "Test PRIV_TOP vs PRIV_TASK"

varnish v1 -vcl {
        import debug;

        backend bad { .host = "${bad_backend}"; }

        sub vcl_recv {
                return (synth(200));
        }

        sub vcl_synth {
                set resp.http.priv_task = debug.test_priv_task("task");
                set resp.http.priv_top = debug.test_priv_top("top");
        }
} -start


client c1 {
        txreq -url
        rxresp
        expect resp.http.priv_task == task
        expect resp.http.priv_top == top
} -run
```

The last `expect` statement fails:

```
**   c1    0.5 === expect resp.http.priv_task == task
**** c1    0.5 EXPECT resp.http.priv_task (task) == "task" match
**   c1    0.5 === expect resp.http.priv_top == top
---- c1    0.5 EXPECT resp.http.priv_top (task) == "top" failed
```

Swapping the two statements inside `vcl_synth` makes `vmod-debug` panic, which is a less friendly way to exhibit the priv conflict. It boils down to both using the (so to speak) top `req` address as the `priv` id in this specific scenario.

A possible solution could be to resurrect `struct reqtop`, and add a field that would hold the `priv_top` id or make the workspace-allocated struct address the id itself, similar to what is done with `req` for client `priv_task`.

I'm well aware why it was retired (d67993451bda7eeb7c0ea3c5528a8353f283a190) but it may have been a necessary evil all along.

**Comment by Dridi** [comment]
By the way, I was able to confirm that back-porting e8985f22711fe9a342b27a10052ae3655f64435c to 6.0 with the merge conflict in `VRT_priv_top` resolved as below is enough to fix this bug.

```diff
diff --git a/bin/varnishd/cache/cache_vrt_priv.c b/bin/varnishd/cache/cache_vrt_priv.c
index 52fd156b7..55802769a 100644
--- a/bin/varnishd/cache/cache_vrt_priv.c
+++ b/bin/varnishd/cache/cache_vrt_priv.c
@@ -152,18 +152,25 @@ VRT_priv_top(VRT_CTX, const void *vmod_id)
 {
        uintptr_t id;
        struct vrt_privs *vps;
+       struct req *req;
 
        CHECK_OBJ_NOTNULL(ctx, VRT_CTX_MAGIC);
-       if (ctx->req) {
-               CHECK_OBJ_NOTNULL(ctx->req, REQ_MAGIC);
-               CHECK_OBJ_NOTNULL(ctx->req->top, REQ_MAGIC);
-               id = (uintptr_t)&ctx->req->top->top;
-               CAST_OBJ_NOTNULL(vps, ctx->req->top->privs, VRT_PRIVS_MAGIC);
-               return (vrt_priv_dynamic(ctx->vcl, ctx->req->top->ws,
-                                        vps, id, (uintptr_t)vmod_id));
-       } else
+       if (ctx->req == NULL) {
+               /* XXX: VRT_fail instead? */
                WRONG("PRIV_TOP is only accessible in client VCL context");
-       NEEDLESS(return NULL);
+               NEEDLESS(return NULL);
+       }
+       CHECK_OBJ_NOTNULL(ctx->req, REQ_MAGIC);
+       Req_MakeTop(ctx->req);
+       if (ctx->req->top == NULL) {
+               VRT_fail(ctx, "PRIV_TOP: out of workspace");
+               return (NULL);
+       }
+       req = ctx->req->top->topreq;
+       id = (uintptr_t)req->top;
+       CAST_OBJ_NOTNULL(vps, req->privs, VRT_PRIVS_MAGIC);
+       return (vrt_priv_dynamic(ctx->vcl, req->ws, vps, id,
+           (uintptr_t)vmod_id));
 }
```

**Comment by Dridi** [comment]
I was initially on the wrong track and reverted the commit that introduced the regression.

**Comment by Dridi** [comment]
Considering how we deal with dynamic priv ids, I don't think we need further attention in this area when it comes to regressions. When I reverted 86af5ce095a6e39e313cf7c6c23a49edba4aace6 I had to bring `(struct vrt_priv).id` back from the grave (retired in 8125596d417bb2629ae146a953862b04ebf55bcd) but the `vcl` field that went away at the same time shouldn't be needed for correctness because dynamic privs are bound to objects themselves bound to a given VCL.

@nigoroll suggested separate trees for top and task scopes of dynamic privs but that's a separate discussion we can open anytime if someone wishes to. As far as fixing the regression I'm convinced I'm done here.
